qubes/storage/lvm.py

```diff
@@ -0,0 +1,344 @@
+        qubes_lvm(cmd, self.log)
+
+    def verify(self, volume):
+        ''' Verifies the volume. '''
+        cmd = ['sudo', 'qubes-lvm', 'volumes',
+               self.volume_group + '/' + self.thin_pool]
+        p = subprocess.Popen(cmd, stdout=subprocess.PIPE)
+        result = p.communicate()[0]
+        for line in result.splitlines():
+            if not line.strip():
+                continue
+            vid, atr = line.strip().split(' ')
+            if vid == volume.vid:
+                return atr[4] == 'a'
+
+        return False
+
+    @property
+    def volumes(self):
+        ''' Return a list of volumes managed by this pool '''
+        cmd = ['sudo', 'qubes-lvm', 'volumes',
+               self.volume_group + '/' + self.thin_pool]
+        p = subprocess.Popen(cmd, stdout=subprocess.PIPE)
+        result = p.communicate()[0]
+        volumes = []
+        for line in result.splitlines():
+            if not line.strip():
+                continue
+            vid, atr = line.strip().split(' ')
+            config = {
+                'pool': self.name,
+                'vid': vid,
+                'name': vid,
+                'volume_group': self.volume_group,
+                'rw': atr[1] == 'w',
+            }
+            volumes += [ThinVolume(**config)]
+        return volumes
+
+    def _reset_volume(self, volume):
+        ''' Resets a volatile volume '''
+        assert volume.volume_type == 'volatile', \
+            'Expected a volatile volume, but got {!r}'.format(volume)
+        self.log.debug('Resetting volatile ' + volume.vid)
+        cmd = ['remove', volume.vid]
+        qubes_lvm(cmd, self.log)
+        cmd = ['create', self._pool_id, volume.vid.split('/')[1],
+               str(volume.size)]
+        qubes_lvm(cmd, self.log)
+
+class ThinVolume(qubes.storage.Volume):
+    ''' Default LVM thin volume implementation
+    '''  # pylint: disable=too-few-public-methods
+
+    def __init__(self, volume_group, **kwargs):
+        self.volume_group = volume_group
+        super(ThinVolume, self).__init__(**kwargs)
+
+        if self.snap_on_start and self.source is None:
+            msg = "snap_on_start specified on {!r} but no volume source set"
+            msg = msg.format(self.name)
+            raise qubes.storage.StoragePoolException(msg)
+        elif not self.snap_on_start and self.source is not None:
+            msg = "source specified on {!r} but no snap_on_start set"
+            msg = msg.format(self.name)
+            raise qubes.storage.StoragePoolException(msg)
+
+        self.path = '/dev/' + self.vid
+        if not self._is_volatile:
+            self._vid_snap = self.vid + '-snap'
+
+    @property
+    def revisions(self):
+        path = self.path + '-back'
+        if os.path.exists(path):
+            seconds = os.path.getctime(path)
+            iso_date = qubes.storage.isodate(seconds).split('.', 1)[0]
+            return {iso_date: path}
+        return {}
```

Comment:
Currently it's hardcoded that lvm can only have 1 revision (as the file store). It would be easy to add, but currently also no tool displays all the revisions.